File: electron/src/proto/codec.js
'use strict';

const crypto = require('node:crypto');

// Framing, authentication and replay rejection — the twin of pkg/proto/codec.go.
//
// Every packet is authenticated: there is no unauthenticated path. With no
// pre-shared key the key is derived from the cluster name alone, which still
// keeps two clusters sharing a LAN apart but, the name being public, provides no
// secrecy.

const NONCE_LEN = 8;
const TS_LEN = 8;
const MAC_LEN = 32;
const HEADER_LEN = 1 + NONCE_LEN + TS_LEN + MAC_LEN;

const KEY_DOMAIN = 'rezoagwe/wire/v2';
const DEFAULT_CLUSTER = 'rezoagwe';
const DEFAULT_SKEW_MS = 30_000;

/** Why a frame was rejected. Each maps to a counter, so drops are never silent. */
const DecodeError = Object.freeze({
  SHORT_FRAME: 'short_frame',
  BAD_MAC: 'bad_mac',
  CLOCK_SKEW: 'clock_skew',
  REPLAY: 'replay',
  MALFORMED: 'malformed',
});

class DecodeException extends Error {
  constructor(reason) {
    super(reason);
    this.name = 'DecodeException';
    this.reason = reason;
  }
}

/** Turns (psk, cluster) into the per-cluster framing key; matches Go's proto.DeriveKey. */
function deriveKey(psk, cluster) {
  const name = cluster || DEFAULT_CLUSTER;
  const mac = crypto.createHmac('sha256', Buffer.from(psk || '', 'utf8'));
  mac.update(Buffer.from(KEY_DOMAIN, 'utf8'));
  mac.update(Buffer.from([0]));
  mac.update(Buffer.from(name, 'utf8'));
  return mac.digest();
}

/**
 * A short, shareable identity for the framing key. Two nodes that cannot talk
 * compare fingerprints to tell "the key differs" from "the network drops
 * packets", without revealing the key.
 */
function keyFingerprint(psk, cluster) {
  return deriveKey(psk, cluster).subarray(0, 4).toString('hex');
}

class Codec {
  constructor(psk, cluster, opts = {}) {
    this.key = deriveKey(psk, cluster);
    this.skewMs = opts.skewMs || DEFAULT_SKEW_MS;
    this.now = opts.now || Date.now;
    this.seen = new Map();
    this.lastPrune = 0;
  }

  /** Frames and authenticates an already-serialised JSON body. */
  encodeBody(kind, body) {
    const bodyBuf = Buffer.isBuffer(body) ? body : Buffer.from(body, 'utf8');
    const frame = Buffer.alloc(HEADER_LEN + bodyBuf.length);
    frame[0] = kind & 0xff;
    crypto.randomFillSync(frame, 1, NONCE_LEN);
    frame.writeBigUInt64BE(BigInt(Math.floor(this.now() / 1000)), 1 + NONCE_LEN);
    bodyBuf.copy(frame, HEADER_LEN);
    const tag = this.tag(frame[0], frame.subarray(1, 1 + NONCE_LEN),
      frame.subarray(1 + NONCE_LEN, 1 + NONCE_LEN + TS_LEN), bodyBuf);
    tag.copy(frame, 1 + NONCE_LEN + TS_LEN);
    return frame;
  }

  /** Frames a message object, serialising it as JSON. */
  encode(kind, value) {
    return this.encodeBody(kind, JSON.stringify(value));
  }

  /**
   * Authenticates a frame and returns { kind, body }. Frames that fail
   * authentication, sit outside the accepted clock skew, or reuse a nonce throw
   * a DecodeException — the caller drops the packet and counts it.
   */
  decode(frame) {
    if (!Buffer.isBuffer(frame) || frame.length < HEADER_LEN) throw new DecodeException(DecodeError.SHORT_FRAME);
    const kind = frame[0];
    const nonce = frame.subarray(1, 1 + NONCE_LEN);
    const tsBytes = frame.subarray(1 + NONCE_LEN, 1 + NONCE_LEN + TS_LEN);
    const mac = frame.subarray(1 + NONCE_LEN + TS_LEN, HEADER_LEN);
    const body = frame.subarray(HEADER_LEN);

    const expected = this.tag(kind, nonce, tsBytes, body);
    if (!crypto.timingSafeEqual(mac, expected)) throw new DecodeException(DecodeError.BAD_MAC);

    const ts = Number(tsBytes.readBigUInt64BE(0)) * 1000;
    const now = this.now();
    const delta = now - ts;
    if (delta > this.skewMs || delta < -this.skewMs) throw new DecodeException(DecodeError.CLOCK_SKEW);

    if (!this.remember(nonce.toString('hex'), now)) throw new DecodeException(DecodeError.REPLAY);
    return { kind, body };
  }

  /** Decodes a frame and parses its JSON body in one step. */
  decodeJSON(frame) {
    const { kind, body } = this.decode(frame);
    let parsed;
    try {
      parsed = JSON.parse(body.toString('utf8'));
    } catch (e) {
      throw new DecodeException(DecodeError.MALFORMED);
    }
    return { kind, body: parsed };
  }

  tag(kind, nonce, ts, body) {
    const mac = crypto.createHmac('sha256', this.key);
    mac.update(Buffer.from([kind]));
    mac.update(nonce);
    mac.update(ts);
    mac.update(body);
    return mac.digest();
  }

  /**
   * Records a nonce and reports whether it was new. Entries older than twice the
   * skew can never be accepted again on timestamp grounds, so they are pruned
   * rather than kept forever.
   */
  remember(nonce, now) {
    if (now - this.lastPrune > this.skewMs) {
      const cutoff = now - 2 * this.skewMs;
      for (const [k, seenAt] of this.seen) if (seenAt < cutoff) this.seen.delete(k);
      this.lastPrune = now;
    }
    if (this.seen.has(nonce)) return false;
    this.seen.set(nonce, now);
    return true;
  }
}

module.exports = {
  Codec,
  DecodeError,
  DecodeException,
  deriveKey,
  keyFingerprint,
  NONCE_LEN,
  TS_LEN,
  MAC_LEN,
  HEADER_LEN,
  DEFAULT_CLUSTER,
  DEFAULT_SKEW_MS,
};

File: pkg/proto/codec.go
package proto

import (
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"encoding/binary"
	"encoding/hex"
	"encoding/json"
	"errors"
	"sync"
	"time"
)

// Framing constants. Every packet is authenticated: there is no unauthenticated
// path to keep tested. When no pre-shared key is configured the key is derived
// from the cluster name alone, which still separates two clusters sharing a
// LAN but — the name being public — provides no secrecy. Pass a -psk for that.
const (
	nonceLen  = 8
	tsLen     = 8
	macLen    = sha256.Size
	headerLen = 1 + nonceLen + tsLen + macLen

	// keyDomain separates this key derivation from any other use of the psk.
	keyDomain = "rezoagwe/wire/v2"

	// DefaultCluster labels a cluster that did not pick a name.
	DefaultCluster = "rezoagwe"

	// DefaultSkew bounds how far a packet's timestamp may sit from local time.
	// It has to cover real clock drift between nodes and the time a packet
	// spends in flight, while staying short enough that the replay window is
	// small.
	DefaultSkew = 30 * time.Second
)

var (
	ErrShortFrame = errors.New("frame shorter than header")
	ErrBadMAC     = errors.New("authentication failed")
	ErrClockSkew  = errors.New("timestamp outside accepted skew")
	ErrReplay     = errors.New("nonce already seen")
)

// DeriveKey turns (psk, cluster) into the per-cluster framing key. An empty psk
// is allowed and yields a cluster-separation-only key.
func DeriveKey(psk, cluster string) []byte {
	if cluster == "" {
		cluster = DefaultCluster
	}
	mac := hmac.New(sha256.New, []byte(psk))
	mac.Write([]byte(keyDomain))
	mac.Write([]byte{0})
	mac.Write([]byte(cluster))
	return mac.Sum(nil)
}

// Codec frames, authenticates and replay-checks packets.
//
// Decode is safe for concurrent use; the replay cache is the only shared state.
type Codec struct {
	key  []byte
	skew time.Duration

	// now is overridable so tests can drive clock-skew and replay-expiry paths
	// without sleeping.
	now func() time.Time

	mu        sync.Mutex
	seen      map[[nonceLen]byte]time.Time
	lastPrune time.Time
}

func NewCodec(psk, cluster string) *Codec {
	return &Codec{
		key:  DeriveKey(psk, cluster),
		skew: DefaultSkew,
		now:  time.Now,
		seen: make(map[[nonceLen]byte]time.Time),
	}
}

// KeyFingerprint is a short, non-secret label for the derived framing key.
//
// Two nodes that cannot talk usually differ in the psk or the cluster name, and
// neither is printable — one is a secret, the other looks right until you
// compare them character by character. A fingerprint is comparable at a glance
// and gives nothing away.
func (c *Codec) KeyFingerprint() string {
	sum := sha256.Sum256(c.key)
	return hex.EncodeToString(sum[:4])
}

// Encode marshals v as the body of a framed, authenticated packet.
func (c *Codec) Encode(kind MessageKind, v interface{}) ([]byte, error) {
	body, err := json.Marshal(v)
	if err != nil {
		return nil, err
	}
	out := make([]byte, headerLen+len(body))
	out[0] = byte(kind)
	if _, err := rand.Read(out[1 : 1+nonceLen]); err != nil {
		return nil, err
	}
	binary.BigEndian.PutUint64(out[1+nonceLen:], uint64(c.now().Unix()))
	copy(out[headerLen:], body)
	tag := c.tag(out[0], out[1:1+nonceLen], out[1+nonceLen:1+nonceLen+tsLen], body)
	copy(out[1+nonceLen+tsLen:], tag)
	return out, nil
}

// Decode authenticates a frame and returns its kind and body. Frames that fail
// authentication, sit outside the accepted clock skew, or reuse a nonce are
// rejected — the caller drops the packet and counts it.
func (c *Codec) Decode(frame []byte) (MessageKind, []byte, error) {
	if len(frame) < headerLen {
		return 0, nil, ErrShortFrame
	}
	kind := MessageKind(frame[0])
	nonce := frame[1 : 1+nonceLen]
	tsBytes := frame[1+nonceLen : 1+nonceLen+tsLen]
	mac := frame[1+nonceLen+tsLen : headerLen]
	body := frame[headerLen:]

	if !hmac.Equal(mac, c.tag(frame[0], nonce, tsBytes, body)) {
		return kind, nil, ErrBadMAC
	}

	ts := time.Unix(int64(binary.BigEndian.Uint64(tsBytes)), 0)
	now := c.now()
	if delta := now.Sub(ts); delta > c.skew || delta < -c.skew {
		return kind, nil, ErrClockSkew
	}

	var key [nonceLen]byte
	copy(key[:], nonce)
	if !c.remember(key, now) {
		return kind, nil, ErrReplay
	}
	return kind, body, nil
}

// DecodeInto decodes a frame and unmarshals its body in one step.
func (c *Codec) DecodeInto(frame []byte, v interface{}) (MessageKind, error) {
	kind, body, err := c.Decode(frame)
	if err != nil {
		return kind, err
	}
	return kind, json.Unmarshal(body, v)
}

func (c *Codec) tag(kind byte, nonce, ts, body []byte) []byte {
	mac := hmac.New(sha256.New, c.key)
	mac.Write([]byte{kind})
	mac.Write(nonce)
	mac.Write(ts)
	mac.Write(body)
	return mac.Sum(nil)
}

// remember records a nonce and reports whether it was new. Entries older than
// twice the skew can never be accepted again on timestamp grounds, so they are
// pruned rather than kept forever.
func (c *Codec) remember(nonce [nonceLen]byte, now time.Time) bool {
	c.mu.Lock()
	defer c.mu.Unlock()
	if now.Sub(c.lastPrune) > c.skew {
		cutoff := now.Add(-2 * c.skew)
		for k, seenAt := range c.seen {
			if seenAt.Before(cutoff) {
				delete(c.seen, k)
			}
		}
		c.lastPrune = now
	}
	if _, dup := c.seen[nonce]; dup {
		return false
	}
	c.seen[nonce] = now
	return true
}
